### PR TITLE
Implement feature score for linear model.

### PR DESCRIPTION
**Pull request by trivialfis**
* Port R feature importance to the new implementation.
* Add linear model support in C++ and Python.

### CODE REVIEW

File: python-package/xgboost/core.py [hunk 1]
```diff
@@ -2132,47 +2132,18 @@ def get_dump(self, fmap='', with_stats=False, dump_format="text"):
         fmap = os.fspath(os.path.expanduser(fmap))
         length = c_bst_ulong()
         sarr = ctypes.POINTER(ctypes.c_char_p)()
-        if self.feature_names is not None and fmap == '':
```
Comment (trivialfis) [inline python-package/xgboost/core.py:2135]: Note: Removed as this is now handled inside C API.

File: jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/BoosterImplTest.java [hunk 2]
```diff
@@ -32,6 +32,9 @@
  * @author hzx
  */
 public class BoosterImplTest {
+  private String train_uri = "../../demo/data/agaricus.txt.train?indexing_mode=1";
```
Comment (trivialfis) [inline jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/java/BoosterImplTest.java:35]: Note: Added `indexing_mode`.

File: src/gbm/gblinear.cc [hunk 3]
```diff
@@ -219,6 +221,24 @@ class GBLinear : public GradientBooster {
     return model_.DumpModel(fmap, with_stats, format);
   }
 
+  void FeatureScore(std::string const &importance_type,
```
Comment (RAMitchell) [inline src/gbm/gblinear.cc:224]: Was the bias term ignored before? I don't want to break existing behaviour, but this term is informative.

**Review by RAMitchell — APPROVED** [review]
LGTM, I've not looked at Java or R code.

File: src/gbm/gblinear.cc (hunk 3 above)
Comment (trivialfis) [reply src/gbm/gblinear.cc:224]: It wasn't included before in R.  This is new for Python as it wasn't supported before.

**Comment by trivialfis** [comment]
Java code should be fine, I'm just fixing some tests.

File: src/gbm/gblinear.cc [hunk 4]
```diff
@@ -219,6 +221,24 @@ class GBLinear : public GradientBooster {
...
+                    std::vector<bst_feature_t> *out_features,
+                    std::vector<float> *out_scores) const override {
+    CHECK(!model_.weight.empty()) << "Model is not initialized";
+    CHECK_EQ(importance_type, "weight")
+        << "gblinear only has `weight` defined for feature importance.";
+    out_features->resize(this->learner_model_param_->num_feature, 0);
+    std::iota(out_features->begin(), out_features->end(), 0);
+    out_scores->resize(model_.weight.size() - learner_model_param_->num_output_group, 0);
```
Comment (hcho3) [inline src/gbm/gblinear.cc:232]:
```suggestion
    // Don't include the bias term in the feature importance scores
    // The bias is the last weight
    out_scores->resize(model_.weight.size() - learner_model_param_->num_output_group, 0);
```

File: R-package/src/xgboost_R.cc [hunk 5]
```diff
@@ -38,34 +38,34 @@
 
 using namespace dmlc;
 
-SEXP XGCheckNullPtr_R(SEXP handle) {
+XGB_DLL SEXP XGCheckNullPtr_R(SEXP handle) {
```
Comment (hcho3) [inline R-package/src/xgboost_R.cc:41]: Why is `XGB_DLL` needed here? Windows users have been using the R package and I don't recall this being an issue.

File: tests/python/test_with_sklearn.py [hunk 6]
```diff
@@ -241,6 +242,35 @@ def test_feature_importances_weight():
                                   importance_type="weight").fit(X, y)
     np.testing.assert_almost_equal(xgb_model.feature_importances_, exp)
 
+    with pytest.raises(ValueError):
+        xgb_model.set_params(importance_type="foo")
+        xgb_model.feature_importances_
+
+    digits = load_digits(n_class=3)
+    y = digits["target"]
+    X = digits["data"]
```
Comment (hcho3) [inline tests/python/test_with_sklearn.py:251]:
```suggestion
    X, y = load_digits(n_class=3, return_X_y=True)
```

**Review by hcho3 — COMMENTED** [review]
LGTM. I left some minor comments.

File: R-package/src/xgboost_R.cc (hunk 5 above)
Comment (trivialfis) [reply R-package/src/xgboost_R.cc:41]: It's already in the header, just matching it so the compiler can check function signature

**Pull request merged**